### ISSUE TITLE
Nap as a flux-like store

### ISSUE THREAD

**Issue by cristiano-belloni** [websdk/nap#14]
This is the other side of #13: I'm using [redux.js](http://rackt.github.io/redux/index.html) at home and I was wondering how to make a flux-like store with Nap / resourceful.

Turns out, IMHO, that they are conceptually very similar:
- Redux has [actions](http://rackt.github.io/redux/docs/basics/Actions.html). An action is "the only source of information for a store", and "must have a type property that indicates the type of action being performed". Sounds similar? Our RESTful requests are actions, with `type` set to one of the HTTP verbs and the rest of the data as a payload.
- Redux has [reducers](http://rackt.github.io/redux/docs/basics/Reducers.html) and [stores](http://rackt.github.io/redux/docs/basics/Store.html). Reducers "specify how the application’s state changes in response". They are essentially request handlers. The store holds state, and can be easily achieved within a module that uses [immutables](https://github.com/websdk/nap/issues/12).
- Redux has [middlewares](http://rackt.github.io/redux/docs/advanced/Middleware.html), which come [out of the box](https://github.com/websdk/nap/blob/master/examples/middleware.html) with Nap.
- The only thing we don't have currently is [subscribing to a store](http://rackt.github.io/redux/docs/basics/Store.html), aka being called back when the store changes. We had the stream type implementation in am-address that would be perfectly suited for it and [probably could have a space in Nap](https://github.com/websdk/nap/issues/5).

**Comment by aaronhaines** [comment]
Interesting. Worth spiking a POC...

**Comment by mstade** [comment]
I'm going to provide a more reasoned response in due time, but I just wanted to quickly comment on your last point about streams. I haven't seen the am-address implementation, so please excuse anything I might say that's redundant, I'll have a look for it later. Anyway, streams is a thing we discussed _a lot_ back in the early days – and whether to make it core or not. Thing is, the request/response model isn't by itself very well suited to streams, and moreover there are different kinds of streams that make finding a generic abstraction difficult. So what we decided back then was that the request/response model stays put, but that you should be able to respond with a stream. This would allow us (and others) to implement different kinds of streams (such as [event source](http://www.w3.org/TR/eventsource/)) while not limiting ourselves. So in essence, the stream isn't a primitive construct like responses. Instead, it _is_ the response – or well, the response _body_ anyhow.

**Comment by cristiano-belloni** [comment]
Yep, understood. Makes sense, and that's how we do it now with zap streams. I guess that, in the spike, we just need to respond with a dispatchers or whatever form of one-way evented communication.

Also, probably #5 is related to this: if we can specify our custom media types, it'll be easy to have a whatever.stream type and use it requesting from our "store".

**Comment by mstade** [comment]
Well, #5 is more to the point of abstracting away hypermedia affordances from the underlying media type.  So HTML for instance has `<form>`, and JSON-inspired hypermedia format may use URI templates or some other construct, but from a Javascript perspective they can be abstracted away as simply functions. There was some work to that effect in the experimental branch I think (and if not, I probably have it not-pushed somewhere.) Frankly, it's more of a client side aspect, and ought to perhaps be part of a pure client project like am-address instead. It's just that I think there may be an opportunity for a feature in nap there to provide for a away to add media type handlers (or "renderers", in the sense that they render a function) such that clients can be composed instead of being monoliths.

**Comment by mstade** [comment]
So, to clarify, the point of adding that level of abstraction is such that you can consume a resource and make use of its hypermedia affordances, without ever really having to know or care what media type the returned representation is. As well, because nap has request metadata (a one-up on actions) even media types that do not support hypermedia (such as plain ol' JSON) could do so via `Link` headers. A consumer wouldn't, and shouldn't, have to care how the links are represented, if nap abstracts that away.

**Comment by mstade** [comment]
And again to clarify, because clearly I can't get it all in one post. You're absolutely on point with the streams thing. :o)
